src/varname.jl

```diff
@@ -383,93 +393,144 @@ function combine_indices(lens::ComposedLens{<:IndexLens})
+    subsumes_indices(left_indices::Tuple, right_indices::Tuple)
 
-Return `true` if `right_index` is subsumed by `left_index`.
+Return `true` if `right_indices` is subsumed by `left_indices`.  `left_indices` is assumed to be 
+concretized and consist of either `Int`s or `AbstractArray`s of scalar indices that are supported 
+by array A.
 
 Currently _not_ supported are: 
 - Boolean indexing, literal `CartesianIndex` (these could be added, though)
 - Linear indexing of multidimensional arrays: `x[4]` does not subsume `x[2, 2]` for a matrix `x`
 - Trailing ones: `x[2, 1]` does not subsume `x[2]` for a vector `x`
-- Dynamic indexing, e.g. `x[1]` does not subsume `x[begin]`.
 """
-subsumes_index(::Tuple{}, ::Tuple{}) = true  # x subsumes x
-subsumes_index(::Tuple{}, ::Tuple) = true    # x subsumes x[1]
-subsumes_index(::Tuple, ::Tuple{}) = false   # x[1] does not subsume x
-function subsumes_index(t::Tuple, u::Tuple)  # does x[i]... subsume x[j]...?
-    return _issubindex(first(t), first(u)) && subsumes_index(Base.tail(t), Base.tail(u))
+subsumes_indices(::Tuple{}, ::Tuple{}) = true  # x subsumes x
+subsumes_indices(::Tuple{}, ::Tuple) = true    # x subsumes x...
+subsumes_indices(::Tuple, ::Tuple{}) = false   # x... does not subsume x
+function subsumes_indices(t1::Tuple, t2::Tuple)  # does x[i]... subsume x[j]...?
+    first_subsumed = all(Base.splat(subsumes_index), zip(first(t1), first(t2)))
+    return first_subsumed && subsumes_indices(Base.tail(t1), Base.tail(t2))
 end
 
-const AnyIndex = Union{Int,AbstractVector{Int},Colon}
-_issubindex_(::Tuple{Vararg{AnyIndex}}, ::Tuple{Vararg{AnyIndex}}) = false
-function _issubindex(t::NTuple{N,AnyIndex}, u::NTuple{N,AnyIndex}) where {N}
-    return all(_issubrange(j, i) for (i, j) in zip(t, u))
+subsumes_index(i::Colon, ::Colon) = error("Colons cannot be subsumed")
+subsumes_index(i, ::Colon) = error("Colons cannot be subsumed")
+subsumes_index(i::Colon, j) = true
+subsumes_index(i::AbstractVector, j) = issubset(j, i)
+subsumes_index(i, j) = i == j
+
+
+struct ConcretizedSlice{T, R} <: AbstractVector{T}
+    range::R
 end
 
-const ConcreteIndex = Union{Int,AbstractVector{Int}} # this include all kinds of ranges
+ConcretizedSlice(s::Base.Slice{R}) where {R} = ConcretizedSlice{eltype(s.indices), R}(s.indices)
+Base.show(io::IO, s::ConcretizedSlice) = print(io, ":")
+Base.show(io::IO, ::MIME"text/plain", s::ConcretizedSlice) = print(io, "ConcretizedSlice($(s.range))")
+Base.IteratorEltype(::Type{<:ConcretizedSlice}) = Base.HasEltype()
+Base.eltype(s::ConcretizedSlice{T}) where {T} = T
+Base.IteratorSize(::Type{<:ConcretizedSlice}) = Base.HasLength()
+Base.length(s::ConcretizedSlice) = length(s.range)
+Base.size(s::ConcretizedSlice) = size(s.range)
+Base.iterate(s::ConcretizedSlice, state...) = Base.iterate(s.range, state...)
+Base.collect(s::ConcretizedSlice) = collect(s.range)
+Base.hasfastin(::Type{<:ConcretizedSlice}) = true
+Base.in(i, s::ConcretizedSlice) = i in s.range
```

Comment:
Most of these definitions should be provided by the defaults for `AbstractVector` automatically. E.g., I don't think it's necessary to define `eltype`, `IteratorEltype`, and `IteratorSize`, and `length` is defined automatically by `size`.